File: src/main/java/com/github/gianttreelp/statsdb/EventListener.java
package com.github.gianttreelp.statsdb;

import com.google.common.collect.Queues;
import org.bukkit.Material;
import org.bukkit.Statistic;
import org.bukkit.entity.EntityType;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.player.PlayerJoinEvent;
import org.bukkit.event.player.PlayerQuitEvent;
import org.bukkit.event.player.PlayerStatisticIncrementEvent;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.Arrays;
import java.util.Objects;
import java.util.Optional;
import java.util.Queue;

public class EventListener implements Listener {

    Queue<StatisticsObject> statisticsQueue = Queues.newConcurrentLinkedQueue();

    @EventHandler
    public void onStatisticIncrement(PlayerStatisticIncrementEvent event) {
        byte[] uuid = StatsDB.getBytesFromUUID(event.getPlayer());

        Optional<StatisticsObject> statObject = statisticsQueue.stream()
                .filter(stat ->
                        Arrays.equals(stat.uuid, uuid) &&
                                event.getStatistic() != null && Objects.equals(stat.statistic, event.getStatistic().name()) &&
                                event.getMaterial() != null && Objects.equals(stat.material, event.getMaterial().name()) &&
                                event.getEntityType() != null && Objects.equals(stat.entity, event.getEntityType().name()))
                .findFirst();

        if (statObject.isPresent()) {
            statObject.get().value = event.getNewValue();
        } else {
            statisticsQueue.add(new StatisticsObject(
                    uuid,
                    event.getStatistic().name(),
                    event.getMaterial() != null ? event.getMaterial().name() : null,
                    event.getEntityType() != null ? event.getEntityType().name() : null,
                    event.getNewValue()
            ));
        }
    }

    @EventHandler
    public void onJoin(PlayerJoinEvent event) {
        byte[] uuid = StatsDB.getBytesFromUUID(event.getPlayer());
        try {
            Connection connection = StatsDB.getConnection();
            PreparedStatement statement = connection.prepareStatement(
                    "SELECT * FROM `Statistics`" +
                            "WHERE `player_id` = ?"
            );
            statement.setBytes(1, uuid);
            ResultSet rs = statement.executeQuery();
            while (rs.next()) {
                Statistic statistic = Statistic.valueOf(rs.getString("statistic"));
                switch (statistic.getType()) {
                    case UNTYPED:
                        event.getPlayer().setStatistic(statistic, rs.getInt("value"));
                        break;
                    case ENTITY:
                        EntityType entity = EntityType.valueOf(rs.getString("entity"));
                        event.getPlayer().setStatistic(statistic, entity, rs.getInt("value"));
                        break;
                    case BLOCK:
                    case ITEM:
                        Material material = Material.valueOf(rs.getString("material"));
                        event.getPlayer().setStatistic(statistic, material, rs.getInt("value"));
                        break;
                }
            }
            rs.close();
            connection.close();
        } catch (SQLException e) {
            e.printStackTrace();
        }
        StatsDB.STATISTIC_LIST.forEach(stat -> statisticsQueue.add(new StatisticsObject(
                uuid,
                stat.name(),
                null,
                null,
                event.getPlayer().getStatistic(stat)
        )));
    }

    @EventHandler
    public void onLeave(PlayerQuitEvent event) {
        byte[] uuid = StatsDB.getBytesFromUUID(event.getPlayer());
        StatsDB.STATISTIC_LIST.forEach(stat -> statisticsQueue.add(new StatisticsObject(
                uuid,
                stat.name(),
                null,
                null,
                event.getPlayer().getStatistic(stat)
        )));
    }
}

File: src/main/kotlin/gtlp/groundmc/statsdb/EventListener.kt
package gtlp.groundmc.statsdb

import com.google.common.collect.Queues
import kotlinx.coroutines.experimental.async
import org.bukkit.Material
import org.bukkit.Statistic
import org.bukkit.entity.EntityType
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.player.PlayerJoinEvent
import org.bukkit.event.player.PlayerQuitEvent
import org.bukkit.event.player.PlayerStatisticIncrementEvent
import java.sql.SQLException
import java.util.*

internal class EventListener : Listener {

    val statisticsQueue: Queue<StatisticsObject> = Queues.newConcurrentLinkedQueue()

    @EventHandler
    fun onStatisticIncrement(event: PlayerStatisticIncrementEvent) {
        val uuid = StatsDB.getBytesFromUUID(event.player)

        val statObject = statisticsQueue.stream()
                .filter { (uuid1, statistic, material, entity) ->
                    Arrays.equals(uuid1, uuid) &&
                            event.statistic !=
                            null && statistic == event.statistic &&
                            event.material != null && material == event.material.name &&
                            event.entityType != null && entity == event.entityType.name
                }
                .findFirst()

        if (statObject.isPresent) {
            statObject.get().value = event.newValue
        } else {
            statisticsQueue.add(StatisticsObject(
                    uuid,
                    event.statistic,
                    if (event.material != null) event.material.name else null,
                    if (event.entityType != null) event.entityType.name else null,
                    event.newValue
            ))
        }
    }

    private val selectStatistics by lazy {
        StatsDB.connection.prepareStatement(
                "SELECT * FROM `Statistics` WHERE `player_id` = ?")
    }

    @EventHandler
    fun onJoin(event: PlayerJoinEvent) {
        async {
            val uuid = StatsDB.getBytesFromUUID(event.player)
            try {
                selectStatistics.setBytes(1, uuid)
                val rs = selectStatistics.executeQuery()
                while (rs.next()) {
                    val statistic = Statistic.valueOf(rs.getString("statistic"))
                    when (statistic.type) {
                        Statistic.Type.UNTYPED -> event.player.setStatistic(statistic, rs.getInt("value"))
                        Statistic.Type.ENTITY -> {
                            val entity = EntityType.valueOf(rs.getString("entity"))
                            event.player.setStatistic(statistic, entity, rs.getInt("value"))
                        }
                        Statistic.Type.BLOCK, Statistic.Type.ITEM -> {
                            val material = Material.valueOf(rs.getString("material"))
                            event.player.setStatistic(statistic, material, rs.getInt("value"))
                        }
                        null -> {
                        }
                    }
                }
                rs.close()
            } catch (e: SQLException) {
                e.printStackTrace()
            }

            StatsDB.STATISTIC_LIST.forEach { stat ->
                statisticsQueue.add(StatisticsObject(
                        uuid,
                        stat, null, null,
                        event.player.getStatistic(stat)
                ))
            }
        }
    }

    @EventHandler
    fun onLeave(event: PlayerQuitEvent) {
        async {
            val uuid = StatsDB.getBytesFromUUID(event.player)
            StatsDB.STATISTIC_LIST.forEach { stat ->
                statisticsQueue.add(StatisticsObject(
                        uuid,
                        stat, null, null,
                        event.player.getStatistic(stat)
                ))
            }
        }
    }
}
